File: Top Interview Questions/L84. 柱状图中最大的矩形.js
/**
 * @param {number[]} heights
 * @return {number}
 */
 var largestRectangleArea = function (heights) {
    let stack = [], max = 0
    heights.push(0)
    heights.unshift(0)
    stack.push([0, heights[0]])
    for (let i = 1; i < heights.length; i++) {
        while (stack.length > 1 && heights[i] < stack.at(-1)[1]) {
            let h=stack.pop()[1]
            let res = h * (i - stack.at(-1)[0]-1)
            max = Math.max(max, res)
        }
        stack.push([i, heights[i]])
    }

    return max
};

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L84柱状图中最大的矩形.java
package top.dyw.leetcode.Hot100;

import java.util.*;
public class L84柱状图中最大的矩形 {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        Deque<Integer> stack = new ArrayDeque<>();
        int[] newHeights = new int[n+2];
        System.arraycopy(heights, 0, newHeights, 1, n);
        int result = 0;
        for (int i=0; i<newHeights.length; i++) {
            while (!stack.isEmpty() && newHeights[stack.peek()] > newHeights[i]) {
                int preIndex = stack.pop();
                int height = newHeights[preIndex];

                int leftIndex = stack.peek();
                int width = i - leftIndex - 1;

                result = Math.max(result, height * width);
            }
            stack.push(i);
        }
        return result;
    }
}
